google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,451 @@
+                INVALID_STATE_ERROR.format("add_response_metadata", self.state)
+            )
+        if self.cluster_id is None or self.zone is None:
+            # BIGTABLE_METADATA_KEY should give a binary-encoded ResponseParams proto
+            blob = cast(bytes, metadata.get(BIGTABLE_METADATA_KEY))
+            if blob:
+                parse_result = self._parse_response_metadata_blob(blob)
+                if parse_result is not None:
+                    cluster, zone = parse_result
+                    if cluster:
+                        self.cluster_id = cluster
+                    if zone:
+                        self.zone = zone
+                else:
+                    self._handle_error(
+                        f"Failed to decode {BIGTABLE_METADATA_KEY} metadata: {blob!r}"
+                    )
+        # SERVER_TIMING_METADATA_KEY should give a string with the server-latency headers
+        timing_header = cast(str, metadata.get(SERVER_TIMING_METADATA_KEY))
+        if timing_header:
+            timing_data = SERVER_TIMING_REGEX.match(timing_header)
+            if timing_data and self.active_attempt:
+                gfe_latency_ms = float(timing_data.group(1))
+                self.active_attempt.gfe_latency_ns = int(gfe_latency_ms * 1e6)
+
+    @staticmethod
+    @lru_cache(maxsize=32)
+    def _parse_response_metadata_blob(blob: bytes) -> Tuple[str, str] | None:
+        """
+        Parse the response metadata blob and return a tuple of cluster and zone.
+
+        Function is cached to avoid parsing the same blob multiple times.
+
+        Args:
+          - blob: the metadata blob as extracted from the grpc call
+        Returns:
+          - a tuple of cluster_id and zone, or None if parsing failed
+        """
+        try:
+            proto = ResponseParams.pb().FromString(blob)
+            return proto.cluster_id, proto.zone_id
+        except (DecodeError, TypeError):
+            # failed to parse metadata
+            return None
+
+    def end_attempt_with_status(self, status: StatusCode | BaseException) -> None:
+        """
+        Called to mark the end of an attempt for the operation.
+
+        Typically, this is used to mark a retryable error. If a retry will not
+        be attempted, `end_with_status` or `end_with_success` should be used
+        to finalize the operation along with the attempt.
+
+        Assumes operation is in ACTIVE_ATTEMPT state.
+
+        Args:
+          - status: The status of the attempt.
+        """
+        if self.state != OperationState.ACTIVE_ATTEMPT or self.active_attempt is None:
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("end_attempt_with_status", self.state)
+            )
+        if isinstance(status, BaseException):
+            status = self._exc_to_status(status)
+        complete_attempt = CompletedAttemptMetric(
+            duration_ns=time.monotonic_ns() - self.active_attempt.start_time_ns,
```

Comment:
should we add a sanity check to make sure this is >= 0? (in case there's bug in the code, we don't want csm to break client code)

Reply:
Ok, I think in java otel if we do otel.record(negative_number) it'll throw some exception. I think we can default to 0 and log a warning